### PR TITLE
USDZ Exporter

### PR DESCRIPTION
**Pull request by mvaligursky**
This PR implements the USDZ export functionality. It allows a hierarchy of entities, along with render components containing meshes, to be exported into USDZ format. This is USDA version under the hood (text format), which is larger, but can be written easily, without using the python based SDK.

Huge thanks to TreeJS for the initial implementation here: https://github.com/mrdoob/three.js/blob/dev/examples/jsm/exporters/USDZExporter.js

Features:
- mesh instances from render component. Meshes are exported with 2 UV channels.
- standard materials with many textures and other properties
- textures stored in png format

The exported file can be inspected by renaming it to .zip file.

Possible improvements: https://github.com/playcanvas/engine/issues/4644

Engine example: USDZ Export, displays a bench glb file, which can be exported to USDZ format. When the example is executed in iPhone Safari, there's an AR icon near the top. When this clicked, the USDZ file is automatically displayed in Quick Look application inside AR.

Screenshot 2022-09-16 at 13 44 52

bench

### CODE REVIEW

File: rollup.config.js [hunk 1]
```diff
@@ -262,7 +262,7 @@ function scriptTarget(name, input, output) {
             file: output || input.replace('.mjs', '.js'),
             format: 'umd',
             indent: '\t',
-            name: name
+            name: name,
```
Comment (willeastcott) [inline rollup.config.js:265]: Remove trailing comma.

File: extras/exporters/usdz-exporter.js [hunk 2]
```diff
@@ -0,0 +1,534 @@
+import { zipSync, strToU8 } from '../../node_modules/fflate/esm/browser.js';
+
+const ROOT_FILE_NAME = 'root';
+
+const header =
+`#usda 1.0
+(
+    customLayerData = {
+        string creator = "Playcanvas UsdzExporter"
```
Comment (willeastcott) [inline extras/exporters/usdz-exporter.js:9]:
```suggestion
        string creator = "PlayCanvas UsdzExporter"
```

File: rollup.config.js (hunk 1 above)
Comment (mvaligursky) [reply rollup.config.js]: yep done thanks

File: examples/src/examples/loaders/usdz-export.tsx [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+            // convert the loaded entity into asdz file
+            new pcx.UsdzExporter().build(entity).then((arrayBuffer: any) => {
+                const blob = new Blob([arrayBuffer], { type: 'application/octet-stream' } );
+
+                // On iPhone Safari, this link creates a clickable AR link on the screen. When this is clicked,
+                // the download of the .asdz file triggers its opening in QuickLook AT mode.
+                // In other browsers, this simply downloads the generated .asdz file.
+
+                // @ts-ignore
+                link.download="bench.usdz";
```
Comment (willeastcott) [inline examples/src/examples/loaders/usdz-export.tsx:66]:
```suggestion
                link.download = "bench.usdz";
```

File: extras/exporters/usdz-exporter.js [hunk 4]
```diff
@@ -0,0 +1,534 @@
...
+
+    /**
+     * A set of used node names. Used in order to keep them unique.
+     *
+     * @type {Set<string>}
+     */
+    nodeNames;
+
+    /**
+     * An object, storing a mapping between the file name and a its content. Used as input to fflate to
```
Comment (willeastcott) [inline extras/exporters/usdz-exporter.js:94]:
```suggestion
     * An object, storing a mapping between the file name and its content. Used as input to fflate to
```

File: extras/exporters/usdz-exporter.js [hunk 5]
```diff
@@ -0,0 +1,534 @@
...
+    // format: https://graphics.pixar.com/usd/release/spec_usdpreviewsurface.html
+    buildMaterial(material) {
+
+        const materialName = `Material_${material.id}`;
+        const materialPath = `/Materials/${materialName}`;
+        const materialPropertyPath = property => `<${materialPath}${property}>`;
+
+        const buildTexture = (texture, textureIds, mapType, uvChannel, tiling, offset, rotation) => {
+
+            // TODO: texture transform values are passed in but do not work corretly in many cases
```
Comment (willeastcott) [inline extras/exporters/usdz-exporter.js:347]:
```suggestion
            // TODO: texture transform values are passed in but do not work correctly in many cases
```

File: package.json [hunk 6]
```diff
@@ -87,7 +87,8 @@
     "@rollup/pluginutils": "^4.2.1",
     "c8": "^7.12.0",
     "chai": "^4.3.6",
-    "eslint": "^8.23.1",
+    "eslint": "^8.23.0",
```
Comment (willeastcott) [inline package.json:90]: Why have you bumped ESLint down?

**Review by willeastcott — APPROVED** [review]
Awesome PR! Approved with some minor feedback. 🚀

File: package.json (hunk 6 above)
Comment (mvaligursky) [reply package.json]: fixed. I had conflict there, and messed up the resolve!

**Review by slimbuck — APPROVED** [review]

File: extras/exporters/usdz-exporter.js [hunk 7]
```diff
@@ -0,0 +1,534 @@
+import { zipSync, strToU8 } from '../../node_modules/fflate/esm/browser.js';
```
[the remark this answers is not in the document]
Comment (mvaligursky) [reply extras/exporters/usdz-exporter.js:1]: do we need to support node and browser build? Or just browser.

[the remark this answers is not in the document]
Comment (slimbuck) [reply extras/exporters/usdz-exporter.js:1]: we probably want to support both.

**Pull request merged**